### ISSUE TITLE
Question about how to declare a field delegate that takes in a list and returns a list

### ISSUE THREAD

**Issue by SapphireBrand** [graphql-dotnet/graphql-dotnet#570]
I have an API where a list is given, and I want to return a list. For example, a valid query would be

{items(ids:[101,202]){status}}

with result

```javascript
{
  "data": {
    "items": [
      {
        "status": 35
      },
      {
        "status": 42
      }
    ]
  }
}
```

The ids are required long integers, which I mapped to NonNullGraphType< ListGraphType< NonNullGraphType< IntGraphType>>>>. Then I declared a FieldDelegate that returns a list of my object graph type, as below.

```c#
FieldDelegate<ListGraphType<ItemObjectGraphType>>(
    "items",
    arguments: new QueryArguments(
    new QueryArgument<NonNullGraphType<ListGraphType<NonNullGraphType<IntGraphType>>>> { Name = "ids", Description = "list of ids of the items" }
    ),
    resolve: new Func<ResolveFieldContext, object, object>((context, id) => ItemsResolveFunc(context, id))
);
```

My question concerns the declaration of "resolve" with the template arguments <ResolveFieldContext, object, object>. I haven't been able to find types other than object and object that match the arguments passed in to ItemsResolveFunc.

graphql-dotnet emphasizes strong typing, so I think that I am missing something. How should I declare the delegate?

**Comment by joemcbride** [comment]
What would you say your end goal is?  What does `ItemsResolveFunc` look like?  Typically usage would look more like:

```
resolve: context =>
{
  var ids = context.GetArgument<List<int>>("ids");
  return GetStatus(ids);
}
```

**Comment by SapphireBrand** [comment]
Yes, the implementation is (well, some details are omitted, but basically...)

```c#
context.GetArgument<List<long>>("ids").Select(id => new Item(id))
```

**Comment by joemcbride** [comment]
So what is your end goal?

**Comment by SapphireBrand** [comment]
To select fields from a list of items. (Is that what you mean by "end goal"?)

Or maybe I am just trying to figure out why the resolve argument is null when my field delegate has a list argument...

When the argument to my FieldDelegate is a scalar, then the resolve function arguments are bound to the value of the argument:

```c#
        // {item(id:30000000101){status}}
        private Task<IItem> ItemResolveFunc(ResolveFieldContext context, long o)
        {
            return Task.FromResult(new Item(o));  // o is 30000000101
        }
```

But when the argument is a list, then o is null:

```c#
        // {items(ids:[30000000101,30000000202]){status}}
        private IEnumerable<IItem> ItemsResolveFunc(ResolveFieldContext context, IEnumerable<long> o)
        {
            // o is null when this function is called, so I am pulling the arguments from the context
            return from id in context.GetArgument<List<long>>("ids")
                   select new OrderedItem(id);
        }
```

I assumed that o is null because I don't know how to declare the resolver, but maybe that's not true. So why is o null when it is called

**Comment by joemcbride** [comment]
Behind the scenes, `FieldDelegate` uses [`MethodModelBinderResolver`](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Resolvers/MethodModelBinderResolver.cs).  At present, I'm pretty sure that `GetArgument` doesn't support `IEnumerable<T>`, so I'm guessing that could be part of the problem.  The code in `MethodModelBinderResolver` attempts to match the arguments based on name.  So the name of the argument defined in your `GraphType` needs to match the argument defined in your `ItemsResolveFunc`.

```csharp
private IEnumerable<IItem> ItemsResolveFunc(ResolveFieldContext context, List<long> ids)
{
    return from id in ids
           select new OrderedItem(id);
}
```

**Comment by SapphireBrand** [comment]
Matching by name! Right, that makes sense. And it happens that my resolvers for scalar GraphTypes all used the same name as the argument types, but my list resolver didn't.

OK, so the pathway is:

1. name the delegate arguments to match the QueryArgument names
1. the delegate argument types match what you want to get in and out
1. at runtime the MethodModelBinderResolver matches arguments by name
1. at runtime, if the types of the supplied values do not convert then you get a runtime exception

One clarification, for the sake of folks who find this answer on the Internet in the future. It isn't the implementation function whose name matters; it is the resolve delegate:

```c#
FieldDelegate<ListGraphType<ItemObjectGraphType>>(
    "items",
    arguments: new QueryArguments(
    new QueryArgument<NonNullGraphType<ListGraphType<NonNullGraphType<IntGraphType>>>> { Name = "ids", Description = "list of ids the items" }
    ),

    // The argument whose name must match the QueryArgument is "ids" in the delegate here
    resolve: new Func<ResolveFieldContext, IEnumerable<long>, IEnumerable<IItem>>((context, ids)
        => ItemsResolveFunc(context, ids)) // <-- declared in the line above as taking arguments (ResolveFieldContext, IEnumerable<long>) and returning IEnumerable<IItem>
);
```
